pkg/proxy/middleware.go

```diff
@@ -0,0 +1,255 @@
+	if maxRetries == 0 {
+		return other
+	}
+	if len(optRetryStatusCodes) == 0 {
+		optRetryStatusCodes = []int{http.StatusBadGateway, http.StatusServiceUnavailable, http.StatusGatewayTimeout}
+	}
+	statusCodeIndex := make(map[int]struct{}, len(optRetryStatusCodes))
+	for _, c := range optRetryStatusCodes {
+		statusCodeIndex[c] = struct{}{}
+	}
+	return &RetryMiddleware{
+		nextHandler:      other,
+		maxRetries:       maxRetries,
+		retryStatusCodes: statusCodeIndex,
+		rSource:          rand.New(rand.NewSource(time.Now().UnixNano())),
+	}
+}
+
+type xResponseWriter interface {
+	http.ResponseWriter
+	discardedResponse() bool
+	CapturedStatusCode() int
+}
+type xBodyCapturer interface {
+	io.ReadCloser
+	Capture()
+}
+
+func (r RetryMiddleware) ServeHTTP(writer http.ResponseWriter, request *http.Request) {
+	lazyBody := newLazyBodyCapturer(request.Body)
+	request.Body = lazyBody
+	for i := 0; ; i++ {
+		discardErrResp := i < r.maxRetries && request.Context().Err() == nil
+		capturedResp := newResponseWriterDelegator(writer, r.isRetryableStatusCode, discardErrResp)
+		// call next handler in chain
+		reqClone := request.Clone(request.Context()) // also copies the reference to the lazy body capturer
+		r.nextHandler.ServeHTTP(capturedResp, reqClone)
+
+		if !capturedResp.discardedResponse() || // max retries reached or context error
+			!r.isRetryableStatusCode(capturedResp.CapturedStatusCode()) {
+			break
+		}
+		// setup for retry
+		lazyBody.Capture()
+
+		totalRetries.Inc()
+		// exponential backoff
+		jitter := time.Duration(r.rSource.Intn(50))
+		time.Sleep(time.Millisecond*time.Duration(1<<uint(i)) + jitter)
+	}
+}
+
+func (r RetryMiddleware) isRetryableStatusCode(status int) bool {
+	_, ok := r.retryStatusCodes[status]
+	return ok
+}
+
+var (
+	_ http.Flusher       = &responseWriterDelegator{}
+	_ io.ReaderFrom      = &xResponseWriterDelegator{}
+	_ statusCodeCapturer = &responseWriterDelegator{}
+)
+
+// responseWriterDelegator represents a wrapper around http.ResponseWriter that provides additional
+// functionalities for handling response writing. Depending on the status code and discard settings,
+// the header + content on write is skipped so that it can be re-used on retry.
+type responseWriterDelegator struct {
+	http.ResponseWriter
+	headerBuf   http.Header
+	wroteHeader bool
+	statusCode  int
+	// always writes to responseWriter when false
+	discardErrResp        bool
+	isRetryableStatusCode func(status int) bool
+}
+
+// newResponseWriterDelegator constructor
+func newResponseWriterDelegator(writer http.ResponseWriter, isRetryableStatusCode func(status int) bool, discardErrResp bool) xResponseWriter {
+	d := &responseWriterDelegator{
+		isRetryableStatusCode: isRetryableStatusCode,
+		ResponseWriter:        writer,
+		headerBuf:             make(http.Header),
+		discardErrResp:        discardErrResp, // abort early on timeout, context cancel
+	}
+	if _, ok := writer.(io.ReaderFrom); ok {
```

Comment:
In Lingo when would the writer be an `io.ReaderFrom` and when would it not be?

Reply:
The `response` type in the Go stdlib implements the interface. So should do all decorators to make use of internal or future optimizations. it is used in `io.Copy` for example. I saw a presentation once but I don't have the link anymore. I also double checked with the prometheus [decorator](https://github.com/prometheus/client_golang/blob/main/prometheus/promhttp/delegator.go#L105) to ensure it is still a good practice

Same for `WriteTo` on the `lazyBodyCapturerWriteTo`